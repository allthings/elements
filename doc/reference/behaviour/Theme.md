<!-- 
This is an auto-generated markdown. 
You can change it in "src/behaviour/Theme.js" and run build:docs to update this file.
-->
# Theme

## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|children **(required)**|func|
