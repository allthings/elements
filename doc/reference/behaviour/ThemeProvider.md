<!-- 
This is an auto-generated markdown. 
You can change it in "src/behaviour/ThemeProvider.jsx" and run build:docs to update this file.
-->
# ThemeProvider
All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
The ThemeProvider allows you to define the default colors for most elements.

**Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|root|bool|
|theme|shape|<br>Default: {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}
|children|node|
