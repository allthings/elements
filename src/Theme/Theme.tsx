import React, { ReactNode, useContext } from 'react'
import {
  defaultTheme,
  ITheme,
  ThemeConsumer,
  ThemeContext,
} from '../ThemeProvider'

type Colorizor = (color: string) => string

export const createColorizor = (theme: ITheme): Colorizor => (color: string) =>
  theme[color] || defaultTheme[color] || color

type ThemeChildrenType = ({
  theme,
  colorize,
}: {
  theme: ITheme
  colorize: (color: string) => string
}) => ReactNode

interface IThemeProps {
  readonly children: ThemeChildrenType
}

export const useTheme = () => {
  const theme = useContext(ThemeContext)

  return {
    theme,
    colorize: createColorizor(theme),
  }
}

const Theme = ({ children }: IThemeProps) => (
  <ThemeConsumer>
    {theme =>
      children({
        theme,
        colorize: createColorizor(theme),
      })
    }
  </ThemeConsumer>
)

export default Theme
