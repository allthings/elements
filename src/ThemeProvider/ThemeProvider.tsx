import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme: ITheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  readonly primary: string
  readonly text: string
  readonly secondaryText: string
  readonly titleColor: string
  readonly contrast: string
  readonly warn: string
  readonly disabled: string
  readonly background: string
  readonly textOnBackground: string
  readonly [key: string]: string
}

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = defaultTheme,
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider
export const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer
