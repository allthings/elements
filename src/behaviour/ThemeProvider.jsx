import React from 'react'
import PropTypes from 'prop-types'
import { Broadcast, Subscriber } from 'react-broadcast'
import { ColorPalette } from '@allthings/colors'

export const THEME_CHANNEL = 'theme'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.primary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

/**
* All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
* The ThemeProvider allows you to define the default colors for most elements.
*
* **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
**/
export default class ThemeProvider extends React.PureComponent {
  static propTypes = {
    root: PropTypes.bool,
    theme: PropTypes.shape({
      primary: PropTypes.string,
      text: PropTypes.string,
      secondaryText: PropTypes.string,
      titleColor: PropTypes.string,
      contrast: PropTypes.string,
      warn: PropTypes.string,
      disabled: PropTypes.string,
      background: PropTypes.string,
      textOnBackground: PropTypes.string,
    }),
    children: PropTypes.node,
  }

  static defaultProps = {
    theme: defaultTheme,
  }

  static contextTypes = Subscriber.contextTypes

  render() {
    let root = true
    if (this.context.broadcasts && this.context.broadcasts[THEME_CHANNEL]) {
      root = false
    }
    const { theme, children } = this.props
    return root ? (
      <Broadcast
        channel={THEME_CHANNEL}
        value={{
          ...ThemeProvider.defaultProps.theme,
          ...theme,
        }}
      >
        {children}
      </Broadcast>
    ) : (
      <Subscriber channel={THEME_CHANNEL}>
        {previousTheme => (
          <Broadcast
            channel={THEME_CHANNEL}
            value={{
              ...ThemeProvider.defaultProps.theme,
              ...previousTheme,
              ...theme,
            }}
          >
            {children}
          </Broadcast>
        )}
      </Subscriber>
    )
  }
}

export const ThemeShape = ThemeProvider.propTypes.theme

export const withTheme = (mapThemeToProps, displayName) => WrappedComponent =>
  // eslint-disable-next-line
  class extends React.PureComponent {
    static displayName = displayName || WrappedComponent.displayName
    static component = WrappedComponent

    renderComponent = theme => {
      const props = !mapThemeToProps
        ? { theme }
        : mapThemeToProps(theme, this.props)

      return <WrappedComponent {...this.props} {...props} />
    }

    render() {
      return (
        <Subscriber channel={THEME_CHANNEL}>{this.renderComponent}</Subscriber>
      )
    }
  }
