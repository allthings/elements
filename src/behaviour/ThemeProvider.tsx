import React, { FunctionComponent, createContext } from 'react'
import { ColorPalette } from '@allthings/colors'

export const defaultTheme = {
  primary: ColorPalette.primary.blue,
  text: ColorPalette.text.primary,
  secondaryText: ColorPalette.text.secondary,
  titleColor: ColorPalette.text.primary,
  contrast: ColorPalette.white,
  warn: ColorPalette.red,
  disabled: ColorPalette.grey,
  background: ColorPalette.background.bright,
  textOnBackground: ColorPalette.white,
}

export interface ITheme {
  primary: string
  text: string
  secondaryText: string
  titleColor: string
  contrast: string
  warn: string
  disabled: string
  background: string
  textOnBackground: string
  [key: string]: string
}

const ThemeContext = createContext<ITheme>(defaultTheme)
export const ThemeConsumer = ThemeContext.Consumer

/**
 * All elements support theming by default, and therefore every element must be wrapped inside a ThemeProvider.
 * The ThemeProvider allows you to define the default colors for most elements.
 *
 * **Example**: If you want all you buttons to be red, instead of writing <Button color="red" /> all the time, you might want to set the "primary" color of your theme to red.
 **/
const ThemeProvider: FunctionComponent<{ theme?: Partial<ITheme> }> = ({
  children,
  theme = {},
}) => (
  <ThemeConsumer>
    {contextTheme => (
      <ThemeContext.Provider value={{ ...contextTheme, ...(theme as ITheme) }}>
        {children}
      </ThemeContext.Provider>
    )}
  </ThemeConsumer>
)

export default ThemeProvider

type mapThemeToPropsType = (theme: ITheme, props: any) => any

function createThemeConsumer<T>(
  mapThemeToProps: mapThemeToPropsType,
  WrappedComponent: React.ComponentType<T>,
  props: Omit<T, 'theme'>
) {
  return (theme: ITheme) => {
    const extraProps = !mapThemeToProps
      ? { theme }
      : mapThemeToProps(theme, props)

    return <WrappedComponent {...props} {...extraProps} />
  }
}

export function withTheme(mapThemeToProps: mapThemeToPropsType) {
  return function<T extends { theme: ITheme }>(
    WrappedComponent: React.ComponentType<T>
  ): React.ComponentType<Omit<T, 'theme'>> {
    return function(props: Omit<T, 'theme'>) {
      return (
        <ThemeConsumer>
          {createThemeConsumer<T>(mapThemeToProps, WrappedComponent, props)}
        </ThemeConsumer>
      )
    }
  }
}
