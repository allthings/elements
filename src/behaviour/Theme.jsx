import React from 'react'
import PropTypes from 'prop-types'
import { ThemeConsumer } from './ThemeProvider'

const createColorizor = theme => color => theme[color] || color

const Theme = ({ children }) => (
  <ThemeConsumer>
    {theme =>
      children({
        theme,
        colorize: createColorizor(theme),
      })
    }
  </ThemeConsumer>
)

Theme.propTypes = {
  children: PropTypes.func.isRequired,
}

export default Theme
