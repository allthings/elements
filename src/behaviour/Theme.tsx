import React, { FunctionComponent } from 'react'
import { ThemeConsumer, ITheme, defaultTheme } from './ThemeProvider'

type Colorizor = (color: string) => string

const createColorizor = (theme: ITheme): Colorizor => (color: string) =>
  theme[color] || defaultTheme[color] || color

type ThemeChildrenType = ({
  theme,
  colorize,
}: {
  theme: ITheme
  colorize: (color: string) => string
}) => React.ReactNode

interface IThemeProps {
  children: ThemeChildrenType
}

const Theme: FunctionComponent<IThemeProps> = ({ children }) => (
  <ThemeConsumer>
    {theme =>
      children({
        theme,
        colorize: createColorizor(theme),
      })
    }
  </ThemeConsumer>
)

export default Theme
