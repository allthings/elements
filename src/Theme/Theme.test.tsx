import * as React from 'react'
import Theme, { useTheme } from './Theme'
import ThemeProvider from '../ThemeProvider'
import { render } from '@testing-library/react'

describe('Theme consumer', () => {
  test('Theme component should render correct color', () => {
    const { container } = render(
      <ThemeProvider theme={{ warn: 'thistle' }}>
        <Theme>{({ colorize }) => colorize('warn')}</Theme>
      </ThemeProvider>,
    )
    expect(container).toMatchInlineSnapshot(`
      <div>
        thistle
      </div>
    `)
  })

  test('useTheme hook should provide correct theme', () => {
    const HookedThemedComponent = () => {
      const { theme } = useTheme()

      return <>{JSON.stringify(theme, null, 2)}</>
    }

    const { container } = render(
      <ThemeProvider>
        <ThemeProvider theme={{ primary: 'thistle', text: 'black' }}>
          <ThemeProvider theme={{ text: 'rebeccapurple' }}>
            <HookedThemedComponent />
          </ThemeProvider>
        </ThemeProvider>
      </ThemeProvider>,
    )

    expect(container).toMatchInlineSnapshot(`
      <div>
        {
        "primary": "thistle",
        "text": "rebeccapurple",
        "secondaryText": "#626262",
        "titleColor": "#333333",
        "contrast": "#ffffff",
        "warn": "#e84c3d",
        "disabled": "#95a5a5",
        "background": "#f3f5f7",
        "textOnBackground": "#ffffff"
      }
      </div>
    `)
  })

  test('useTheme hook should provide correct value for `colorize`', () => {
    const HookedThemedComponent = () => {
      const { colorize } = useTheme()

      return <>{JSON.stringify(colorize('warn'))}</>
    }
    const { container } = render(
      <ThemeProvider theme={{ warn: 'thistle' }}>
        <HookedThemedComponent />
      </ThemeProvider>,
    )
    expect(container).toMatchInlineSnapshot(`
      <div>
        "thistle"
      </div>
    `)
  })
})
