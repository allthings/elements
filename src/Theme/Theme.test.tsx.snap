// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correct color 1`] = `"#e84c3d"`;
